### DISCUSSION TITLE
Struggling with Regex

### DISCUSSION THREAD

**Discussion by gitboy16** [vlang/v#14051] [General]
Hi,

I looked at the documentation of the regex module but I am still struggling with it. I was wondering if you would be able to provide examples on how to do the following.

I have the following text:
```
text := 'title="hello John" title="subject 1"'
```
I would like to extract 'hello John' and 'subject 1' using regex.

I have tried the following: 
```
query := r"(title=[\w.]+)+"
```

Thank you

**Comment by larpon** [comment]
Your  regex doesn't account for the double quotes?
Try using a literal raw V string:
`query := r'title="(\w+)"'` (untested)

**Comment by gitboy16** [comment]
Thank you. To account for the space in word I changed it to :
```
query := r'title="(\w.+)"'
```
However I only get the first occurence. Any idea how to get the other one(s) as well?
Thank you

**Reply by larpon** [reply]
You should be able to iterate capture groups as [explained in the "Groups" section here](https://modules.vlang.io/regex.html#readme_regex)
... you could also try using [`find_all_str`](https://modules.vlang.io/regex.html#RE.find_all_str)

**Comment by gitboy16** [comment]
The problem as well with that query is that is caught `title="hello John"` instead of catching `hello John`.

**Reply by larpon** [reply]
That sounds like a potential bug since the capture group is defined as `(\w.+)` - if it persists you could open an issue so we can get it fixed.

**Reply by gitboy16** [reply]
Please check on your side because it is indeed the case and hence may be a vug.

**Reply by larpon** [reply]
It depends on how you're calling/using the regex.
Can you provide a minimal piece of code so we can reproduce the issue?

**Reply by gitboy16** [reply]
Here is some code:

```vlang
import regex
fn main(){
 txt := 'title="ww w" title="ci1 ao"'
query := r'title="(\w.+)"'
mut re := regex.regex_opt(query) or { panic(err) } 
start, end := re.match_string(txt)
if start >= 0 {
println("Match ($start, $end) => [${txt[start..end]}]") 
} else {
 println("No Match")
 }
mat := re.find_all_str(txt)
println(mat)
 }
```
In both cases what is captured is `title="<words>"` and not the actual `<words>`.

**Reply by larpon** [reply]
Thanks

**Reply by larpon** [reply]
@penguindark does this look like a bug to you?

**Reply by penguindark** [reply]
@gitboy16 in this case the regex capture `title="<words>"` because you are matching the string, and the match start at 0.
If you want only the word you can simply get only the first group like in this code:
```v
import regex
fn main(){
    txt := 'title="ww w" test words'
    query := r'title="(\w.+)"'
    mut re := regex.regex_opt(query) or { panic(err) } 
    start, end := re.match_string(txt)
    if start >= 0 {
        println("Match ($start, $end) => [${txt[start..end]}]")
        println("Group 1 => [${re.get_group_by_id(txt,0)}]")
    } else {
        println("No Match")
    }
}
```
Remeber that the match and the groups are different things and that you have in memory only the last group matched.

If you want only the contained text in a "continuos" check you can use the https://github.com/vlang/v/tree/master/vlib/regex#groups-continuous-saving

Or you can use a simply and more fast (because it doesn't use groups):
```v
import regex
fn main(){
    txt := 'title="ww w" test words title="pero" melo title="bastd" orco'
    query := r'title="\w.+"'
    mut re := regex.regex_opt(query) or { panic(err) }
    mut res := re.find_all_str(txt).map(it#[7..-1])
    println(res)
}
```

But remeber that this simple tasks can be done with simple strings manipualtion.

**Reply by gitboy16** [reply]
Thank you all. Kind regards.

**Comment by Wajinn** [comment]
> text := 'title="hello John" title="subject 1"'

Just wanted to add that simpler string manipulation in V shouldn't always be overlooked.

```
text := 'title="hello John" title="subject 1"'

for value in text.split('title=') {
	if value !='' {
		println(value)
	}
}

for value in text.split('"') {
	if value !='' && !value.contains('title') {
		println(value)
	}
}
```
